### PR TITLE
gridplot controllers kwarg is back, other improvements to gp

### LINKED ISSUE #425: Directly pass `Controller` instances to `GridPlot`
#382 brought in one regression, we can no longer directly pass a controller instance, this is useful for synchronizing across plots so we should bring it back. For now users can manually assign controller instances for subplots _after_ creating a `GridPlot`.

### PR DESCRIPTION
**Pull request by kushalkolar**
closes #425

also adds tests the various cameras and controllers kwargs, and changing cameras and controllers after gridplot create

### CODE REVIEW

File: fastplotlib/layouts/_gridplot.py [hunk 1]
```diff
@@ -110,127 +117,177 @@ def __init__(
...
+                _controller = create_controller(
+                    controller_type=cont_type, camera=cams[0]
+                )
+
+                subplot_controllers[controller_ids == cid] = _controller
+
+                # add the other cameras that go with this controller
+                if cams.size > 1:
+                    for cam in cams[1:]:
+                        _controller.add_camera(cam)
```
Comment (kushalkolar) [inline fastplotlib/layouts/_gridplot.py:290]: This entire block isn't new, it's now in the `else` block if `controllers is None`

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -145,6 +146,7 @@ jobs:
       env:
         PYGFX_EXPECT_LAVAPIPE: true
       run: |
+        WGPU_FORCE_OFFSCREEN=1 pytest -v tests/
```
Comment (kushalkolar) [inline .github/workflows/ci.yml:149]: we can make this more elegant later

File: fastplotlib/layouts/_gridplot.py [hunk 3]
```diff
@@ -21,16 +21,21 @@ def __init__(
...
+            | Iterable[Literal["panzoom", "fly", "trackball", "orbit"]]
         ) = None,
-        controller_ids: str | list[int] | np.ndarray | list[list[str]] = None,
+        controller_ids: (
+            Literal["sync"]
+            | Iterable[int]
+            | Iterable[Iterable[int]]
+            | Iterable[Iterable[str]]
+        ) = None,
+        controllers: pygfx.Controller | Iterable[Iterable[pygfx.Controller]] = None,
```
Comment (kushalkolar) [inline fastplotlib/layouts/_gridplot.py:38]: just some better typing

File: fastplotlib/layouts/_gridplot.py [hunk 4]
```diff
@@ -110,127 +117,177 @@ def __init__(
...
+            # one controller for all subplots
+            if isinstance(controllers, pygfx.Controller):
+                controllers = [controllers] * len(self)
+                # subplot_controllers[:] = controllers
+                # # subplot_controllers = np.asarray([controllers] * len(self), dtype=object)
 
-        elif isinstance(controller_ids, str):
-            if controller_ids == "sync":
-                controller_ids = np.zeros(self.shape, dtype=int)
+            # individual controller instance specified for each subplot
```
Comment (kushalkolar) [inline fastplotlib/layouts/_gridplot.py:133]: if controller instances are manually specified

File: fastplotlib/layouts/_gridplot.py [hunk 5]
```diff
@@ -282,6 +344,31 @@ def renderer(self) -> pygfx.WgpuRenderer:
...
+        cameras.flags.writeable = False
+        return cameras
+
+    @property
+    def names(self) -> np.ndarray[str]:
+        """subplot names, read-only array, access individual subplots to change a name"""
+        names = np.asarray([subplot.name for subplot in self]).reshape(self.shape)
+        names.flags.writeable = False
+        return names
+
```
Comment (kushalkolar) [inline fastplotlib/layouts/_gridplot.py:371]: new

File: tests/test_gridplot.py
Comment (kushalkolar) [inline tests/test_gridplot.py:1]: @clewis7 any more tests you can think of?

File: fastplotlib/layouts/_gridplot.py [hunk 6]
```diff
@@ -70,6 +74,11 @@ def __init__(
             | list of lists of subplot names, each sublist is synced: [[subplot_a, subplot_b, subplot_e], [subplot_c, subplot_d]]
             | this syncs subplot_a, subplot_b and subplot_e together; syncs subplot_c and subplot_d together
 
+        controllers: pygfx.Controller | list[pygfx.Controller] | np.ndarray[pygfx.Controller], optional
+            directly provide pygfx.Controller instances(s). Useful if you want to use a controller from an existing
+            plot/subplot. Other controller kwargs, i.e. ``controller_Types`` and ``controller_ids`` are ignored if
```
Comment (clewis7) [inline fastplotlib/layouts/_gridplot.py:79]: small typo here 

``controller_types``

File: fastplotlib/layouts/_gridplot.py [hunk 7]
```diff
@@ -110,127 +117,177 @@ def __init__(
...
+                # result we want for arrays
+                for item in controllers:
+                    if isinstance(item, pygfx.Controller):
+                        pass
+                    elif all(isinstance(c, pygfx.Controller) for c in item):
+                        pass
+                    else:
+                        raise TypeError(
+                            "controllers argument must be a single pygfx.Controller instance of a Iterable of "
+                            "pygfx.Controller instances"
```
Comment (clewis7) [inline fastplotlib/layouts/_gridplot.py:145]: "controllers argument must be a single pygfx.Controller instance or an Iterable of
 pygfx.Controller instances"

File: fastplotlib/layouts/_gridplot.py [hunk 8]
```diff
@@ -110,127 +117,177 @@ def __init__(
...
+                        )
 
-        # list controller_ids
-        elif isinstance(controller_ids, (list, np.ndarray)):
-            ids_flat = list(chain(*controller_ids))
+            try:
+                controllers = np.asarray(controllers).reshape(shape)
+            except ValueError:
+                raise ValueError(
```
Comment (clewis7) [inline fastplotlib/layouts/_gridplot.py:151]: Is this just to raise our own unique `ValueError` lol

Comment (clewis7) [reply fastplotlib/layouts/_gridplot.py:151]: seems wonky

Comment (kushalkolar) [reply fastplotlib/layouts/_gridplot.py:151]: https://github.com/fastplotlib/fastplotlib/pull/432#discussion_r1523564380

**Pull request merged**